qiskit/algorithms/gradients/spsa_sampler_gradient.py

```diff
@@ -0,0 +1,123 @@
+
+        Raises:
+            ValueError: If ``epsilon`` is not positive.
+        """
+        if epsilon <= 0:
+            raise ValueError(f"epsilon ({epsilon}) should be positive.")
+        self._batch_size = batch_size
+        self._epsilon = epsilon
+        self._seed = np.random.default_rng(seed) if seed else np.random.default_rng()
+
+        super().__init__(sampler, **run_options)
+
+    def _run(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        parameter_values: Sequence[Sequence[float]],
+        parameters: Sequence[Sequence[Parameter] | None],
+        **run_options,
+    ) -> SamplerGradientResult:
+        """Compute the sampler gradients on the given circuits."""
+        jobs, offsets, metadata_ = [], [], []
+        for circuit, parameter_values_, parameters_ in zip(circuits, parameter_values, parameters):
+            # indices of parameters to be differentiated
+            if parameters_ is None:
+                indices = list(range(circuit.num_parameters))
+            else:
+                indices = [circuit.parameters.data.index(p) for p in parameters_]
+            metadata_.append({"parameters": [circuit.parameters[idx] for idx in indices]})
+
+            offset = np.array(
+                [
+                    (-1) ** (self._seed.integers(0, 2, len(circuit.parameters)))
+                    for _ in range(self._batch_size)
+                ]
+            )
+            plus = [parameter_values_ + self._epsilon * offset_ for offset_ in offset]
+            minus = [parameter_values_ - self._epsilon * offset_ for offset_ in offset]
+            offsets.append(offset)
+
+            job = self._sampler.run([circuit] * 2 * self._batch_size, plus + minus, **run_options)
+            jobs.append(job)
+
+        # combine the results
+        if any(job.status() is not JobStatus.DONE for job in jobs):
+            raise QiskitError("The gradient job was not completed successfully. ")
+        results = [job.result() for job in jobs]
+        gradients = []
+        for i, result in enumerate(results):
+            grad_dists = np.zeros((self._batch_size, 2 ** circuits[i].num_qubits))
+            for j, (dist_plus, dist_minus) in enumerate(
+                zip(result.quasi_dists[: self._batch_size], result.quasi_dists[self._batch_size :])
+            ):
+                grad_dists[j, list(dist_plus.keys())] += list(dist_plus.values())
+                grad_dists[j, list(dist_minus.keys())] -= list(dist_minus.values())
+            grad_dists /= 2 * self._epsilon
+            gradient_ = []
+            indices = [circuits[i].parameters.data.index(p) for p in metadata_[i]["parameters"]]
+            for j in range(circuits[i].num_parameters):
+                if not j in indices:
+                    continue
+                grad = np.mean(
+                    np.array([delta * dist for dist, delta in zip(grad_dists, offsets[i][:, j])]),
```

Comment:
Not a blocker, but it would be nice to be consistent with variable names and dividing by the offset like in the estimator gradients 🙂